include/drivers/uart.h

```diff
@@ -1460,6 +1192,319 @@ static inline void uart_irq_callback_set(const struct device *dev,
 	uart_irq_callback_user_data_set(dev, cb, NULL);
 }
 
+/**
+ * @}
+ */
+
+/**
+ * @addtogroup uart_async
+ * @{
+ */
+
+/**
+ * @brief Set event handler function.
+ *
+ * Since it is mandatory to set callback to use other asynchronous functions,
+ * it can be used to detect if the device supports asynchronous API. Remaining
+ * API does not have that detection.
+ *
+ * @param dev       UART device structure.
+ * @param callback  Event handler.
+ * @param user_data Data to pass to event handler function.
+ *
+ * @retval -ENOSYS  If not supported by the device.
+ * @retval -ENOTSUP If API not enabled.
+ * @retval 0	    If successful, negative errno code otherwise.
```

Comment:
nit: mixes tabs/spaces. In general I prefer this format (easier to maintain):

```suggestion
 * @retval 0 If successful, negative errno code otherwise.
```

But we have no rules really, so up to you.